# Import libraries

In [13]:
import numpy as np

import pandas as pd

import matplotlib.pyplot as plt

import seaborn as sns

import sklearn

# read all 12 csv files into a dataframe

In [14]:
# ## keith's method
# # import os library

# # store the csv files in a list. the csvs's directory path as the parameter for .listdir()
# import os           
# files = [file for file in os.listdir('sales data')]         

# # create an empty dataframe to store the csvs later
# df =pd.DataFrame()

# # iterate over each file in files and concatenate with df above
# for file in files:
#     # read the file as a dataframe
#     month = pd.read_csv(f'sales data\{file}')

#     # concatenate with the empty dataframe, df
#     df = pd.concat([df, month], axis = 0)

# df

df1 = pd.read_csv(r'sales data\Sales_January_2019.csv')
df2 = pd.read_csv(r'sales data\Sales_February_2019.csv')
df3 = pd.read_csv(r'sales data\Sales_March_2019.csv')
df4 = pd.read_csv(r'sales data\Sales_April_2019.csv')
df5 = pd.read_csv(r'sales data\Sales_May_2019.csv')
df6 = pd.read_csv(r'sales data\Sales_June_2019.csv')
df7 = pd.read_csv(r'sales data\Sales_July_2019.csv')
df8 = pd.read_csv(r'sales data\Sales_August_2019.csv')
df9 = pd.read_csv(r'sales data\Sales_September_2019.csv')
df10 = pd.read_csv(r'sales data\Sales_October_2019.csv')
df11 = pd.read_csv(r'sales data\Sales_November_2019.csv')
df12 = pd.read_csv(r'sales data\Sales_December_2019.csv')

df1

Order ID                   Product Quantity Ordered Price Each  \
0      141234                    iPhone                1        700   
1      141235  Lightning Charging Cable                1      14.95   
2      141236          Wired Headphones                2      11.99   
3      141237          27in FHD Monitor                1     149.99   
4      141238          Wired Headphones                1      11.99   
...       ...                       ...              ...        ...   
9718   150497              20in Monitor                1     109.99   
9719   150498          27in FHD Monitor                1     149.99   
9720   150499           ThinkPad Laptop                1     999.99   
9721   150500    AAA Batteries (4-pack)                2       2.99   
9722   150501              Google Phone                1        600   

          Order Date                       Purchase Address  
0     01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1     01/28/19 14:15       185 Maple St, Portland, OR 97035  
2     01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3     01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4     01/25/19 11:59          387 10th St, Austin, TX 73301  
...              ...                                    ...  
9718  01/26/19 19:09            95 8th St, Dallas, TX 75001  
9719  01/10/19 22:58    403 7th St, San Francisco, CA 94016  
9720  01/21/19 14:31        214 Main St, Portland, OR 97035  
9721  01/15/19 14:21      810 2nd St, Los Angeles, CA 90001  
9722  01/13/19 16:43         428 Cedar St, Boston, MA 02215  

[9723 rows x 6 columns]

In [15]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9723 entries, 0 to 9722
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          9697 non-null   object
 1   Product           9697 non-null   object
 2   Quantity Ordered  9697 non-null   object
 3   Price Each        9697 non-null   object
 4   Order Date        9697 non-null   object
 5   Purchase Address  9697 non-null   object
dtypes: object(6)
memory usage: 455.9+ KB


In [16]:
df1.describe()

Order ID               Product Quantity Ordered Price Each  \
count       9697                  9697             9697       9697   
unique      9269                    20                8         19   
top     Order ID  USB-C Charging Cable                1      11.95   
freq          16                  1171             8795       1171   

        Order Date  Purchase Address  
count         9697              9697  
unique        8077              9161  
top     Order Date  Purchase Address  
freq            16                16

# combine all dataframe

In [17]:
df = pd.concat([df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11, df12], axis = 0)
df

Order ID                     Product Quantity Ordered Price Each  \
0       141234                      iPhone                1        700   
1       141235    Lightning Charging Cable                1      14.95   
2       141236            Wired Headphones                2      11.99   
3       141237            27in FHD Monitor                1     149.99   
4       141238            Wired Headphones                1      11.99   
...        ...                         ...              ...        ...   
25112   319666    Lightning Charging Cable                1      14.95   
25113   319667       AA Batteries (4-pack)                2       3.84   
25114   319668             Vareebadd Phone                1        400   
25115   319669            Wired Headphones                1      11.99   
25116   319670  Bose SoundSport Headphones                1      99.99   

           Order Date                        Purchase Address  
0      01/22/19 21:25         944 Walnut St, Boston, MA 02215  
1      01/28/19 14:15        185 Maple St, Portland, OR 97035  
2      01/17/19 13:33   538 Adams St, San Francisco, CA 94016  
3      01/05/19 20:33      738 10th St, Los Angeles, CA 90001  
4      01/25/19 11:59           387 10th St, Austin, TX 73301  
...               ...                                     ...  
25112  12/11/19 20:58  14 Madison St, San Francisco, CA 94016  
25113  12/01/19 12:01    549 Willow St, Los Angeles, CA 90001  
25114  12/09/19 06:43        273 Wilson St, Seattle, WA 98101  
25115  12/03/19 10:39          778 River St, Dallas, TX 75001  
25116  12/21/19 21:45  747 Chestnut St, Los Angeles, CA 90001  

[186850 rows x 6 columns]

# What's the best month for sales? How much was earned that month?

In [18]:
# rename columns
df.columns = ['id', 'product', 'quantity', 'price', 'date', 'address']
df

id                     product quantity   price            date  \
0      141234                      iPhone        1     700  01/22/19 21:25   
1      141235    Lightning Charging Cable        1   14.95  01/28/19 14:15   
2      141236            Wired Headphones        2   11.99  01/17/19 13:33   
3      141237            27in FHD Monitor        1  149.99  01/05/19 20:33   
4      141238            Wired Headphones        1   11.99  01/25/19 11:59   
...       ...                         ...      ...     ...             ...   
25112  319666    Lightning Charging Cable        1   14.95  12/11/19 20:58   
25113  319667       AA Batteries (4-pack)        2    3.84  12/01/19 12:01   
25114  319668             Vareebadd Phone        1     400  12/09/19 06:43   
25115  319669            Wired Headphones        1   11.99  12/03/19 10:39   
25116  319670  Bose SoundSport Headphones        1   99.99  12/21/19 21:45   

                                      address  
0             944 Walnut St, Boston, MA 02215  
1            185 Maple St, Portland, OR 97035  
2       538 Adams St, San Francisco, CA 94016  
3          738 10th St, Los Angeles, CA 90001  
4               387 10th St, Austin, TX 73301  
...                                       ...  
25112  14 Madison St, San Francisco, CA 94016  
25113    549 Willow St, Los Angeles, CA 90001  
25114        273 Wilson St, Seattle, WA 98101  
25115          778 River St, Dallas, TX 75001  
25116  747 Chestnut St, Los Angeles, CA 90001  

[186850 rows x 6 columns]

In [29]:
# remove null rows
df = df.dropna(axis = 0, how = 'any')

# change data type of columns 'quantity' and 'price' to int and float respectively



id          0
product     0
quantity    0
price       0
date        0
address     0
dtype: int64

In [45]:
# get rid of 'Price Each' and 'Quantity Ordered' values in price and quantity column respectively
df = df.loc[(df['price'] != 'Price Each') | (df['quantity'] != 'Quantity Ordered')]
df

id                     product quantity   price            date  \
0      141234                      iPhone        1     700  01/22/19 21:25   
1      141235    Lightning Charging Cable        1   14.95  01/28/19 14:15   
2      141236            Wired Headphones        2   11.99  01/17/19 13:33   
3      141237            27in FHD Monitor        1  149.99  01/05/19 20:33   
4      141238            Wired Headphones        1   11.99  01/25/19 11:59   
...       ...                         ...      ...     ...             ...   
25112  319666    Lightning Charging Cable        1   14.95  12/11/19 20:58   
25113  319667       AA Batteries (4-pack)        2    3.84  12/01/19 12:01   
25114  319668             Vareebadd Phone        1     400  12/09/19 06:43   
25115  319669            Wired Headphones        1   11.99  12/03/19 10:39   
25116  319670  Bose SoundSport Headphones        1   99.99  12/21/19 21:45   

                                      address  
0             944 Walnut St, Boston, MA 02215  
1            185 Maple St, Portland, OR 97035  
2       538 Adams St, San Francisco, CA 94016  
3          738 10th St, Los Angeles, CA 90001  
4               387 10th St, Austin, TX 73301  
...                                       ...  
25112  14 Madison St, San Francisco, CA 94016  
25113    549 Willow St, Los Angeles, CA 90001  
25114        273 Wilson St, Seattle, WA 98101  
25115          778 River St, Dallas, TX 75001  
25116  747 Chestnut St, Los Angeles, CA 90001  

[185950 rows x 6 columns]

In [65]:
# convert values in quantity column to int8
df['quantity'] = pd.to_numeric(df['quantity'], errors = 'coerce').astype(np.int8)

# convert price column to float
df['price'] = pd.to_numeric(df['price'], errors = 'coerce').astype(np.float64)

df

id                     product  quantity   price            date  \
0      141234                      iPhone         1  700.00  01/22/19 21:25   
1      141235    Lightning Charging Cable         1   14.95  01/28/19 14:15   
2      141236            Wired Headphones         2   11.99  01/17/19 13:33   
3      141237            27in FHD Monitor         1  149.99  01/05/19 20:33   
4      141238            Wired Headphones         1   11.99  01/25/19 11:59   
...       ...                         ...       ...     ...             ...   
25112  319666    Lightning Charging Cable         1   14.95  12/11/19 20:58   
25113  319667       AA Batteries (4-pack)         2    3.84  12/01/19 12:01   
25114  319668             Vareebadd Phone         1  400.00  12/09/19 06:43   
25115  319669            Wired Headphones         1   11.99  12/03/19 10:39   
25116  319670  Bose SoundSport Headphones         1   99.99  12/21/19 21:45   

                                      address  
0             944 Walnut St, Boston, MA 02215  
1            185 Maple St, Portland, OR 97035  
2       538 Adams St, San Francisco, CA 94016  
3          738 10th St, Los Angeles, CA 90001  
4               387 10th St, Austin, TX 73301  
...                                       ...  
25112  14 Madison St, San Francisco, CA 94016  
25113    549 Willow St, Los Angeles, CA 90001  
25114        273 Wilson St, Seattle, WA 98101  
25115          778 River St, Dallas, TX 75001  
25116  747 Chestnut St, Los Angeles, CA 90001  

[185950 rows x 6 columns]

In [67]:
# create 'sales' colum
df['sales'] = df['price']*df['quantity']
df

id                     product  quantity   price            date  \
0      141234                      iPhone         1  700.00  01/22/19 21:25   
1      141235    Lightning Charging Cable         1   14.95  01/28/19 14:15   
2      141236            Wired Headphones         2   11.99  01/17/19 13:33   
3      141237            27in FHD Monitor         1  149.99  01/05/19 20:33   
4      141238            Wired Headphones         1   11.99  01/25/19 11:59   
...       ...                         ...       ...     ...             ...   
25112  319666    Lightning Charging Cable         1   14.95  12/11/19 20:58   
25113  319667       AA Batteries (4-pack)         2    3.84  12/01/19 12:01   
25114  319668             Vareebadd Phone         1  400.00  12/09/19 06:43   
25115  319669            Wired Headphones         1   11.99  12/03/19 10:39   
25116  319670  Bose SoundSport Headphones         1   99.99  12/21/19 21:45   

                                      address   sales  
0             944 Walnut St, Boston, MA 02215  700.00  
1            185 Maple St, Portland, OR 97035   14.95  
2       538 Adams St, San Francisco, CA 94016   23.98  
3          738 10th St, Los Angeles, CA 90001  149.99  
4               387 10th St, Austin, TX 73301   11.99  
...                                       ...     ...  
25112  14 Madison St, San Francisco, CA 94016   14.95  
25113    549 Willow St, Los Angeles, CA 90001    7.68  
25114        273 Wilson St, Seattle, WA 98101  400.00  
25115          778 River St, Dallas, TX 75001   11.99  
25116  747 Chestnut St, Los Angeles, CA 90001   99.99  

[185950 rows x 7 columns]

In [68]:
df.describe()

quantity          price          sales
count  185950.000000  185950.000000  185950.000000
mean        1.124383     184.399735     185.490917
std         0.442793     332.731330     332.919771
min         1.000000       2.990000       2.990000
25%         1.000000      11.950000      11.950000
50%         1.000000      14.950000      14.950000
75%         1.000000     150.000000     150.000000
max         9.000000    1700.000000    3400.000000

In [73]:
# convert date column values to datetime64 object
df['date'] = pd.to_datetime(df['date'], format = '%m/%d/%y %H:%M')

0       2019-01-22 21:25:00
1       2019-01-28 14:15:00
2       2019-01-17 13:33:00
3       2019-01-05 20:33:00
4       2019-01-25 11:59:00
                ...        
25112   2019-12-11 20:58:00
25113   2019-12-01 12:01:00
25114   2019-12-09 06:43:00
25115   2019-12-03 10:39:00
25116   2019-12-21 21:45:00
Name: date, Length: 185950, dtype: datetime64[ns]

In [76]:
# create new month column
df['month'] = pd.DatetimeIndex(df['date']).month
df

id                     product  quantity   price            date  \
0      141234                      iPhone         1  700.00  01/22/19 21:25   
1      141235    Lightning Charging Cable         1   14.95  01/28/19 14:15   
2      141236            Wired Headphones         2   11.99  01/17/19 13:33   
3      141237            27in FHD Monitor         1  149.99  01/05/19 20:33   
4      141238            Wired Headphones         1   11.99  01/25/19 11:59   
...       ...                         ...       ...     ...             ...   
25112  319666    Lightning Charging Cable         1   14.95  12/11/19 20:58   
25113  319667       AA Batteries (4-pack)         2    3.84  12/01/19 12:01   
25114  319668             Vareebadd Phone         1  400.00  12/09/19 06:43   
25115  319669            Wired Headphones         1   11.99  12/03/19 10:39   
25116  319670  Bose SoundSport Headphones         1   99.99  12/21/19 21:45   

                                      address   sales  month  
0             944 Walnut St, Boston, MA 02215  700.00      1  
1            185 Maple St, Portland, OR 97035   14.95      1  
2       538 Adams St, San Francisco, CA 94016   23.98      1  
3          738 10th St, Los Angeles, CA 90001  149.99      1  
4               387 10th St, Austin, TX 73301   11.99      1  
...                                       ...     ...    ...  
25112  14 Madison St, San Francisco, CA 94016   14.95     12  
25113    549 Willow St, Los Angeles, CA 90001    7.68     12  
25114        273 Wilson St, Seattle, WA 98101  400.00     12  
25115          778 River St, Dallas, TX 75001   11.99     12  
25116  747 Chestnut St, Los Angeles, CA 90001   99.99     12  

[185950 rows x 8 columns]

In [85]:
# groupby month and then take sum of sales in each month
sales_groupby_month = df.groupby('month')['sales'].sum()
# df.groupby('month')[['sales', 'quantity']].sum()

sales_groupby_month.sort_values(ascending = False)

month
12    4613443.34
10    3736726.88
4     3390670.24
11    3199603.20
5     3152606.75
3     2807100.38
7     2647775.76
6     2577802.26
8     2244467.88
2     2202022.42
9     2097560.13
1     1822256.73
Name: sales, dtype: float64

## So, the best month for sales is in December wih total sales of RM 4,613,443